library/ssl_tls.c

```diff
@@ -4751,10 +4751,10 @@ int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
     return( 0 );
 }
 
-mbedtls_ssl_srtp_profile
-     mbedtls_ssl_get_dtls_srtp_protection_profile( const mbedtls_ssl_context *ssl )
+const mbedtls_dtls_srtp_info *
+     mbedtls_ssl_get_dtls_srtp_negotiation_result( const mbedtls_ssl_context *ssl )
 {
-    return( ssl->dtls_srtp_info.chosen_dtls_srtp_profile );
+    return( &( ssl->dtls_srtp_info ) );
```

Comment:
As far as I understand from the standard a typical use case is to do a handshake to establish an SRTP association, then immediately trigger a DTLS renegotiation several times to establish more SRTP associations for further data streams. In this case the `dtls_srtp_info` in the `ssl` context might be overwritten before the user accesses it. This would result in a silent error that depends on timing and is potentially very hard to debug, especially if there is concurrency involved.

I think it would be more robust to take a `mbedtls_dtls_srtp_info` output parameter as well and copy the negotiated values there.

Also, since we are copying already, we could check if the protection profile is `MBEDTLS_TLS_SRTP_UNSET` and if it is, set/leave `mki_len` to 0.

Reply:
No problem to change but if you're afraid the user won't call the get result function before a renegotiation and thus loose the first result, copying the output won't resolve it as the copy would be performed when the function is called so if a renegotiation is already ongoing we might have lost the result of the first handshake already.

The user could also copy the output content from the current version of the function, but if you feel more secure with copying it in the get function, it's fine for me.

Reply:
If the user doesn't call the function before renegotiation then it can't really be helped. But if the user does call the get result function before a renegotiation, they still might access the returned pointer at some later point and by that time the pointed data will reflect the most recently negotiated protection profile and MKI.

You are right, the user can do the copying themselves, it is just more friendly and robust if we do it.

Reply:
I made the change to return a copied struct and not the pointer to the internal one

Reply:
Thank you for the fast update and for your patience!